README.md
# Mojar::Cron

Cron-style time patterns.


Changes
0.402 2017-05-14
  - Upgrade for Mojolicious v7.15.
  - Now requires Perl v5.14.

0.341 2016-07-15
  - Fix tz_offset (Util).

0.331 2016-06-28
  - Fix tz_offset (Util).

0.322 2016-06-25
  - Fix test (Util).
  - Remove v5.24 from travisci cfg.

0.311 2016-06-20
  - Fix conditional declaration (Cron).

0.301 2016-05-24
  - Add experimental Date class (Date).
  - Fix bug #1 (Datetime).

0.241 2015-01-17
  - Add date_dow (Util).

0.231 2014-09-08
  - Update handling of useragent errors (Holiday::*).
  - Bump Mojolicious req s/3.33/5/.

0.221 2014-09-02
  - Add experimental tz_offset (Util).

0.211 2014-08-25
  - Remove class attributes (Datetime).

0.202 2014-08-21
  - Add loader from gov.uk (Holiday::UkGov).
  - Add loader from kayaposoft.com (Holiday::Kayaposoft).
  - Add date_today (Util).

0.131 2014-05-27
  - Add holiday base (Holiday).

0.121 2014-03-03
  - Add date_next (Util).

0.112 2014-02-13
  - Change to use Mojar::ClassShare (Datetime).
  - Tidy packaging.

0.100 2013-08-03
  - Initial release.
  - Fix bug mapping day names to numbers (Cron).


Makefile.PL
use 5.010_001;
use strict;
use warnings;

use ExtUtils::MakeMaker;

WriteMakefile(
  NAME         => 'Mojar::Cron',
  VERSION_FROM => 'lib/Mojar/Cron.pm',
  ABSTRACT     => 'Cron-style time patterns and algorithm',
  AUTHOR       => 'Nic Sandfield <niczero@cpan.org>',
  LICENSE      => 'artistic_2',
  META_MERGE   => {
    requires  => {perl => 5.010001},
    resources => {
      homepage    => 'https://niczero.github.io/mojar',
      license     => 'http://www.opensource.org/licenses/artistic-license-2.0',
      repository  => 'https://github.com/niczero/mojar-cron',
      bugtracker  => 'https://github.com/niczero/mojar-cron/issues'
    },
    no_index => {directory => [qw(test)]}
  },
  BUILD_REQUIRES => {'Test::More' => 0},
  CONFIGURE_REQUIRES => {'ExtUtils::MakeMaker' => 6.3},
  PREREQ_PM => {
    Mojolicious => 3.33,
    Mojar => 1.053
  },
  test => {TESTS => 'test/*.t'}
);


lib/Mojar/Cron.pm
package Mojar::Cron;
use Mojo::Base -base;

our $VERSION = 0.402;

use Carp 'croak';
use Mojar::Cron::Datetime;
use POSIX qw(mktime strftime setlocale LC_TIME);

# Fields of a cron pattern
our @Fields = qw(sec min hour day month weekday);

# Soft limits for defining ranges
my %Max; @Max{@Fields} = (59, 59, 23, 30, 11, 6);

# Array indices of a datetime record
use constant {
  SEC     => 0,
  MIN     => 1,
  HOUR    => 2,
  DAY     => 3,
  MONTH   => 4,
  YEAR    => 5,
  WEEKDAY => 6
};
# NB the distinction (YEAR) between this and @Fields

# Canonical names
my (%Month, %Weekday);
{
  my $old_locale = setlocale(LC_TIME);
  setlocale(LC_TIME, 'C');

  %Month   = map +(lc(strftime '%b', 0,0,0,1,$_,70) => $_), 0 .. 11;
  %Weekday = map +(lc(strftime '%a', 0,0,0,$_,5,70) => $_), 0 .. 6;

  setlocale(LC_TIME, $old_locale);
}

# Attributes

# Object has these seven attributes
has \@Fields;
has 'is_local';

# Public methods

sub new {
  my ($class, %param) = @_;

  # Exclude is_local from expansion
  my $is_local = delete $param{is_local};

  # Identify time pattern attributes
  my @values;
  if (exists $param{pattern}) {
    @values = split /\s+/, delete $param{pattern};
  }
  elsif (exists $param{parts}) {
    @values = @{ delete $param{parts} };
  }
  else {
    my $given_sec = exists $param{sec} ? 1 : undef;
    @values = map $param{$_}, @Fields;
    $values[0] = 0 unless $given_sec;  # Do not expand sec more than requested
    delete @param{@Fields};
  }
  # Apply default 'sec'
  unshift @values, '0' if @values < 6;  # Vivify sec

  croak(sprintf 'Unrecognised parameter (%s)', join ',', keys %param) if %param;
  %param = ();

  # Expand parameter values
  $param{$Fields[$_]} = expand($Fields[$_] => $values[$_]) for 0 .. 5;
  
  return $class->SUPER::new(%param, is_local => $is_local);
}

sub expand {
  # Function; not method
  my ($field, $spec) = @_;

  return undef if not defined $spec or $spec eq '*';

  my @vals;
  for my $val (split /,/, $spec) {
    my $step = 1;
    my $end;

    $val =~ s|/(\d+)$|| and $step = $1;

    $val =~ /^(.+)-(.+)$/ and ($val, $end) = ($1, $2);

    if ($val eq '*') {
      ($val, $end) = (0, $Max{$field});
    }
    elsif ($field eq 'day') {
      # Externally 1-31; internally 0-30
      defined and /^\d+$/ and --$_ for $val, $end;
    }
    elsif ($field eq 'month') {
      # Externally 1-12; internally 0-11
      defined and /^\d+$/ and --$_ for $val, $end;
      # Convert symbolics
      defined and exists $Month{lc $_} and $_ = $Month{lc $_} for $val, $end;
    }
    elsif ($field eq 'weekday') {
      # Convert symbolics
      defined and exists $Weekday{lc $_} and $_ = $Weekday{lc $_}
        for $val, $end;
      $end = 7 if defined $end and $end == 0 and $val > 0;
    }

    push @vals, $val;
    push @vals, $val while defined $end and ($val += $step) <= $end;

    if ($field eq 'weekday' and $vals[-1] == 7) {
      unshift @vals, 0 unless $vals[0] == 0;
      pop @vals;
    }
  }
  return [ sort {$a <=> $b} @vals ];
}

sub next {
  my ($self, $previous) = @_;
  # Increment to the next possible time and convert to datetime
  my $dt = Mojar::Cron::Datetime->from_timestamp(
      $previous + 1, $self->is_local);

  {
    redo unless $self->satisfiable(MONTH, $dt);

    if (defined $self->{day} and defined $self->{weekday}) {
      # Both day and weekday are defined, so the cron entry should trigger as
      # soon as _either_ of them is satisfied.  Therefore need to determine
      # which is satisfied sooner.

      my $weekday_dt = $dt->new;
      my $weekday_restart = not $self->satisfiable(WEEKDAY, $weekday_dt);
      my $next_by_weekday = $weekday_dt->to_timestamp($self->is_local);

      my $day_dt = $dt->new;
      my $day_restart = not $self->satisfiable(DAY, $day_dt);
      my $next_by_day = $day_dt->to_timestamp($self->is_local);

      if ($next_by_day <= $next_by_weekday) {
        $dt->copy($day_dt);
        redo if $day_restart;
      }
      else {
        $dt->copy($weekday_dt);
        redo if $weekday_restart;
      }
    }
    elsif (defined $self->{day}) {
      redo unless $self->satisfiable(DAY, $dt);
    }
    elsif (defined $self->{weekday}) {
      redo unless $self->satisfiable(WEEKDAY, $dt);
    }

    redo unless $self->satisfiable(HOUR, $dt);
    redo unless $self->satisfiable(MIN, $dt);
    redo unless $self->satisfiable(SEC, $dt);
  }

  return $dt->to_timestamp($self->is_local);
}

sub satisfiable {
  my ($self, $component, $dt) = @_;

  # The given $component of $self is a sequence of numeric slots.  Test whether
  # those slots can satisfy the corresponding component of $dt.  Shortcircuit
  # 'true' if slot is a wildcard.
  my $field = ($component == WEEKDAY) ? 'weekday' : $Fields[$component];
  my $slots = $self->{$field} // return 1;

  # $old : existing value; $new : same or next value satisfying cron
  my $old = ($component == WEEKDAY) ? $dt->weekday : $dt->[$component];
  my $new;
  # Grab first slot at least big enough
  for (@$slots) { $new = $_, last if $_ >= $old }

  # Can't manipulate WEEKDAY directly since it is tied to DAY.  Instead
  # manipulate DAY until goal is achieved.
  if ($component == WEEKDAY) {
    $component = DAY;
    $new = $dt->[DAY] + $new - $old if defined $new;  # adjust by the same delta
    $old = $dt->[DAY];

    if (not defined $new) {
      # Rollover (into following week)
      $dt->reset_parts(DAY - 1);
      # Add more days till we hit next occurrance of $slots->[0]
      # We know $dt->weekday is greater than all @$slots, so the following will
      # add less than a week to $dt->[DAY].
      $dt->[DAY] += $slots->[0] + 7 - $dt->weekday;

      $dt->normalise;
      return undef;
    }
    elsif ($new > $old) {
      # Component has moved up a slot
      $dt->reset_parts($component - 1);
    }
  }
  elsif (not defined $new) {
    # Rollover
    $dt->reset_parts($component - 1);
    $dt->[$component] = $slots->[0];
    $dt->[$component + 1]++;

    $dt->normalise;
    return undef;
  }
  elsif ($new > $old) {
    # Component has moved up a slot
    $dt->reset_parts($component - 1);
  }

  $dt->[$component] = $new;

  # Detect rollover of month and reset to next month
  my $was_month = $dt->[MONTH];
  $dt->normalise;

  if ($component == DAY and $was_month != $dt->[MONTH]) {
    $dt->reset_parts(DAY);
    return undef;
  }

  return 1;
}

1;
__END__

=head1 NAME

Mojar::Cron - Cron-style datetime patterns and algorithm

=head1 SYNOPSIS

  use Mojar::Cron;
  my $c = Mojar::Cron->new(is_local => 1, pattern => '0 9 * * mon');
  my $t = time;
  $t = $c->next($t);  # next instance of 9am Monday
  say "Need to wait another $t seconds."

=head1 DESCRIPTION

A time pattern is a sequence of six components: second, minute, hour, day,
month, weekday (the component for 'second' may be omitted for brevity and
consistency with standard cron syntax) and supports C<@>-expressions in the
weekday component.  Each of the first four components consists of a
comma-separated list of patterns.

  2
  1,3,5
  1-4,6
  */2

The 'month' component can use those patterns, but may also use English prefices
in place of numbers.

  apr-jun,oct-dec  # instead of 4-6,10-12

The 'weekday' component can do similarly.

  mon-fri  # instead of 1-5
  mon,wed,fri  # instead of 1,3,5

(Note that this implementation also supports 'wrap around' patterns such as
C<fri-mon> and C<oct-mar>.)

In keeping with standard cron, a time pattern may specify sequences for both day
and weekday in the same pattern.  This can cause confusion because the pattern
is satisfied if either sequence is satisfied.

  00 00 00 01  * mon  # midnight on the first of each month and also each Monday

The weekday component may use C<@>-expressions, but in this case the 'day' and
'month' components must be unspecified.

  00 00 00  *  * @mon#1      # first Monday of each month
  00 00 16  *  * @5#L        # last Friday of each month
  00 00 16  *  * @week#L     # last week day (Mon-Fri) of each month
  00 00 03  *  * @weekend#1  # first weekend day (Sat-Sun) of each month

There are also special expressions which can replace the whole time pattern.

  00 01 00  *  * *  # @nightly
  00 01 06  *  * *  # @morningly
  00 01 12  *  * *  # @daily
  00 01 18  *  * *  # @eveningly

Cron patterns of the kind C<*/n> have a weakness that they only specify jumps
within their sequence.  For example the 'day' pattern C<*/15> will trigger on
1st, 16th, 31st January and then again on 1st February; the pattern resets at
the end of its sequence (month days).  This might not be what you expect or
want, so we also have yearday patterns.  These have the same behaviour, but
because they don't reset till the end of the year, they get closer to what
people expect.

  00 00 00  *  * :*/2  # every two days through the year

=head1 METHODS

=head2 C<new>

=head2 C<next>

=head2 C<expand>

=head2 C<satisfiable>

=head1 TROUBLESHOOTING

For consistency with cron, patterns are one-based for day and month, but
internally (using Mojar::Cron::Datetime) everything is zero-based.  So
when debugging your code C<[0,0,0,0,0,0]> would be '1900-01-01 00:00:00' and
C<[[0],[0],[0],undef,undef,[1]]> would be '00 00 00 * * 1'.

=head1 COPYRIGHT AND LICENCE

The main algorithm is thanks to Paul Evans (leonerd@leonerd.org.uk).  I hope I
have implemented it in a readable way, and it turns out doing so revealed
several bugs in the original.

Copyright (C) 2012, Paul Evans.

Copyright (C) 2012--2016, Nic Sandfield.

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 SEE ALSO

L<Algorithm::Cron> (Paul has merged in bugfixes I sent.)


lib/Mojar/Cron/Util.pm
package Mojar::Cron::Util;
use Mojo::Base -strict;

our $VERSION = 0.041;

use Carp 'croak';
use Exporter 'import';
use POSIX qw(mktime strftime);
use Time::Local 'timegm';

our @EXPORT_OK = qw(
  time_to_zero zero_to_time cron_to_zero zero_to_cron life_to_zero zero_to_life
  balance normalise_utc normalise_local date_today date_next date_previous
  date_dow utc_to_ts local_to_ts ts_to_utc ts_to_local local_to_utc utc_to_local
  tz_offset
);

# Public functions

sub time_to_zero { @_[0..2], $_[3] - 1, @_[4..$#_] }
sub zero_to_time { @_[0..2], $_[3] + 1, @_[4..$#_] }

sub cron_to_zero { @_[0..2], $_[3] - 1, $_[4] - 1, @_[5..$#_] }
sub zero_to_cron { @_[0..2], $_[3] + 1, $_[4] + 1, @_[5..$#_] }

sub life_to_zero { @_[0..2], $_[3] - 1, $_[4] - 1, $_[5] - 1900, @_[6..$#_] }
sub zero_to_life { @_[0..2], $_[3] + 1, $_[4] + 1, $_[5] + 1900, @_[6..$#_] }

sub balance {
  my @parts = @_;
  my @Max = (59, 59, 23, undef, 11);
  # Bring values within range for sec, min, hour, month (zero-based)
  for (0,1,2,4) {
    $parts[$_] += $Max[$_] + 1, --$parts[$_ + 1] while $parts[$_] < 0;
    $parts[$_] -= $Max[$_] + 1, ++$parts[$_ + 1] while $parts[$_] > $Max[$_];
  }
  return @parts;
}

sub normalise_utc {
  my @parts = balance @_;
  my $days = $parts[3] - 1;  # could be negative
  my $ts = timegm @parts[0..2], 1, @parts[4..$#parts];  # first of the month
  $ts += $days * 24 * 60 * 60;
  return gmtime $ts;
}

sub normalise_local {
  my @parts = balance @_;
  my $days = 0;
  if ($parts[3] < 1 or 28 < $parts[3] && $parts[4] == 1 or 30 < $parts[3]) {
    $days = $parts[3] - 1;  # possibly negative
    $parts[3] = 1;
  }
  my $ts = mktime @parts;
  $ts += $days * 24 * 60 * 60;
  return localtime $ts;
}

sub date_today { strftime '%Y-%m-%d', localtime }

sub date_next {
  strftime '%Y-%m-%d', 0,0,0, $3 + 1, $2 - 1, $1 - 1900
    if shift =~ /^(\d{4})-(\d{2})-(\d{2})\b/;
}

sub date_previous {
  strftime '%Y-%m-%d', 0,0,0, $3 - 1, $2 - 1, $1 - 1900
    if shift =~ /^(\d{4})-(\d{2})-(\d{2})\b/;
}

sub date_dow {
  strftime '%u', 0,0,0, $3 + 1, $2 - 1, $1 - 1900
    if shift =~ /^(\d{4})-(\d{2})-(\d{2})\b/;
}

sub utc_to_ts    { timegm @_ }
sub local_to_ts  { mktime @_ }

sub ts_to_utc    { gmtime $_[0] }
sub ts_to_local  { localtime $_[0] }

sub local_to_utc { gmtime mktime @_ }
sub utc_to_local { localtime timegm @_ }

my %UnitFactor = (
  S => 1,
  M => 60,
  H => 60 * 60,
  d => 60 * 60 * 24,
  w => 60 * 60 * 24 * 7,
  m => 60 * 60 * 24 * 30,
  y => 60 * 60 * 24 * 365
);

sub str_to_delta {
  my ($str) = @_;
  return 0 unless $str;
  return $str if $str =~ /^[-+]?\d+S?$/;
  return $1 * $UnitFactor{$2} if $str =~ /^([-+]?\d+)([MHdwmy])$/;
  croak qq{Failed to interpret time period ($str)};
}

sub tz_offset {
  my $now = shift // time;
  my ($lm, $lh, $ly, $ld) = (localtime $now)[1, 2, 5, 7];
  my ($um, $uh, $uy, $ud) = (gmtime $now)[1, 2, 5, 7];
  my $min = $lm - $um + 60 * ($lh - $uh) + 60 * 24 * ($ly - $uy || $ld - $ud);
  return _format_offset($min);
}

# Private function

# This is simply to aid unit testing
sub _format_offset {
  my $min = shift;
  my $sign = $min < 0 ? '-' : '+';
  my $hr = int(($min + 0.5) / 60);
  $min = $min - 60 * $hr;
  return sprintf '%s%02u%02u', $sign, abs($hr), abs($min);
}

1;
__END__

=head1 NAME

Mojar::Cron::Util - Time utility functions

=head1 SYNOPSIS

  use Mojar::Cron::Util 'date_next';

=head1 DESCRIPTION

Utility functions for dates and times.

=head1 FUNCTIONS

=head2 time_to_zero

  ($S, $M, $H, $d, $m, $y) = time_to_zero($S, $M, $H, $d, $m, $y);

Converts time representations to zero-based datetimes.  So day 1 translates to
0, while months and years are left 0-based.

=head2 zero_to_time

  ($S, $M, $H, $d, $m, $y) = zero_to_time($S, $M, $H, $d, $m, $y);

Converts zero-based datetimes to time representations.  So day 0 translates to
1, but months and years are left 0-based.

=head2 cron_to_zero

  ($S, $M, $H, $d, $m, $y) = cron_to_zero($S, $M, $H, $d, $m, $y);

Converts cron representations to zero-based datetimes.  So day 1 translates to
0, month 1 translates to 0 (January), while years are left 0-based.

=head2 zero_to_cron

  ($S, $M, $H, $d, $m, $y) = zero_to_cron($S, $M, $H, $d, $m, $y);

Converts zero-based datetimes to cron representations.  So day 0 translates to
1, month 0 translates to 1 (January), but years are left 0-based.

=head2 life_to_zero

  ($S, $M, $H, $d, $m, $y) = life_to_zero($S, $M, $H, $d, $m, $y);

Converts real-life representations to zero-based datetimes.  So day 1 translates
to 0, month 1 translates to 0 (January), and year 1900 translates to 0.

=head2 zero_to_life

  ($S, $M, $H, $d, $m, $y) = zero_to_life($S, $M, $H, $d, $m, $y);

Converts zero-based datetimes to real-life representations.  So day 0 translates
to 1, month 0 translates to 1 (January), and year 0 translates to 1900.

=head2 balance

  ($S, $M, $H, $d, $m, $y) = balance($S, $M, $H, $d, $m, $y);

Balance-out any simple-minded anomalies such as seconds being less than 0 or
greater than 59, or days being less than 0 or greater than 31.  This lets you
make crude adjustments, such as adding 30 mins, and then letting this function
balance it back into the realms of normality.  Note that it takes care of
everything except the length of months, and so is mainly only used by the two
normalise functions which will handle that.

=head2 normalise_utc

  ($S, $M, $H, $d, $m, $y) = normalise_utc($S, $M, $H, $d, $m, $y);

Normalises a UTC datetime to a valid value.  For example, 31 April translates to
1 May.

=head2 normalise_local

  ($S, $M, $H, $d, $m, $y) = normalise_local($S, $M, $H, $d, $m, $y);

Normalises a local datetime to a valid value.  For example, 31 April translates
to 1 May.

=head2 date_today

  $today = date_today();  # yyyy-mm-dd

Provides today's date, using the local (system) clock.

=head2 date_previous

  $previous = date_previous('2015-03-01');
  $yesterday = date_previous(date_today());

Provides the previous date.

=head2 date_next

  $next = date_next('2015-02-28');
  $tomorrow = date_next(date_today());

Provides the following date.

=head2 utc_to_ts

=head2 local_to_ts

=head2 ts_to_utc

=head2 ts_to_local

=head2 local_to_utc

=head2 utc_to_local

=head2 str_to_delta

=head2 tz_offset

  $offset = tz_offset;
  $offset = tz_offset($epoch);

Provides the numeric timezone offset, taking daylight saving into account.  It
is more portable than

  POSIX::strftime('%z')

as it works on non-nix platforms such as Windows.  This is required by some date
formats, such as in SMTP.

=head1 COPYRIGHT AND LICENCE

Copyright (C) 2012--2016, Nic Sandfield.

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 SEE ALSO

L<Mojar::Util>, L<POSIX>.


lib/Mojar/Cron/Holiday.pm
package Mojar::Cron::Holiday;
use Mojo::Base -base;

use Carp 'croak';
use Mojar::Cron::Datetime;

# Attributes

has holidays => sub { {} };
has 'linked';

# Methods

sub holiday {
  my ($self, $date) = (shift, shift);
  $date //= substr(Mojar::Cron::Datetime->now->to_string, 0, 10);
  my $holidays = $self->holidays;
  my $linked = $self->linked;

  if (ref $date eq 'ARRAY') {
    # Recurse over a setter bundle
    return [map $self->holiday($_ => $_[0]), @$date] if @_;

    # Recurse over a getter bundle
    return [map $self->holiday($_), @$date];
  }

  # Scalar
  croak 'Bad format' unless $date =~ /^\d\d\d\d-\d\d-\d\d$/;

  # Setter
  $holidays->{$date} = !! $_[0], return $self if @_;

  # Getter
  return $holidays->{$date} if exists $holidays->{$date};

  # Defer if possible
  return $linked->holiday($date) if $linked;

  # Negative
  return undef;
}

sub next_holiday {
  my ($self, $date) = @_;
  $date //= substr(Mojar::Cron::Datetime->now->to_string, 0, 10);
  my $holidays = $self->holidays;

  my @dates = sort grep $holidays->{$_}, keys %$holidays;
  shift @dates while @dates and $dates[0] lt $date;
  @dates ? $dates[0] : undef;
}

1;
__END__

=head1 NAME

Mojar::Cron::Holiday - Cache for the holidays

=head1 SYNOPSIS

  my $national = Mojar::Cron::Holiday->new(holidays => {
    '2014-12-24' => 1,
    '2014-12-25' => 1,
    '2015-01-01' => 1,
  });
  say 'Yippee!' if $national->holiday('2014-12-25');

  my $regional = Mojar::Cron::Holiday->new(holidays => {
    '2014-12-25' => 0,  # We work Christmas Day
    '2014-12-26' => 1   # but get Boxing Day in lieu
  });
  $regional->linked($national);  # regional is an overlay on national
  say 'Booo!' unless $regional->holiday('2014-12-25');
  say 'Phew!'
    if $regional->holiday('2014-12-24') and $regional->holiday('2014-12-26');

=head1 DESCRIPTION

=head1 ATTRIBUTES

=head2 holidays

  $calendar->holidays({'2016-01-01' => 1});

A hashref holding the underlying holiday list.  A false value against a date
indicates it is not a holiday; useful when overriding a linked calendar.

It is equally convenient to set the holiday list using the holidays attribute
(with a hashref) or the holiday method (with an arrayref); choose whichever you
find more readable within your code.

=head2 linked

  $regional->linked($national);
  say 'National holiday' if $regional->linked->holiday('2016-01-04');

The linked attribute may hold another holiday object, to which queries are
passed whenever the current object has no opinion.  Thus a tree of holiday nodes
can be built.  The root of the tree might be national public holidays; below
that might be corporate holidays for the boss's birthday and when the warehouse
shuts for stocktaking; and below that might be a node for the annual leave of
each employee.  Commonly the tree is simpler: a node for project 'down' days,
linking to a node for public holidays.

=head1 METHODS

=head2 holiday

  $calendar->holiday('2016-01-04' => 1);
  $calendar->holiday(['2016-03-25', '2016-03-28'] => 0);
  say 'Bye!' if $calendar->holiday;
  say 'Discard the milk' if $calendar->holiday($yesterday);

A getter and setter for whether this calendar flags the date(s) as holiday.  The
date can be either a scalar (string) or an arrayref of such scalars.  If no date
is given, the current date is used.

When used as a setter, the second argument is a boolean.  For base or standalone
calendars, only positive results need be assigned; negatives are assigned when
you want to override a possible linked result.

=head2 next_holiday

  $anticipated = $calendar->next_holiday('2016-02-01');

Returns the next holiday on or following the given date.  If no date is given,
the current date is used.

=head1 RATIONALE

I need a small web service so that various apps can forecast when staff will be
in better moods.  The prime example is a gantt editor, which relies on holiday
information for forecasting when there is a glimmer of hope of work being done.

=head1 COPYRIGHT AND LICENCE

Copyright (C) 2014, Nic Sandfield.

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.


lib/Mojar/Cron/Timestamp.pm
package Mojar::Cron::Timestamp;
use Mojo::Base -base;

use Carp 'carp';
use POSIX qw( mktime strftime );
use Time::Local 'timegm';

# Class attributes

has format => '%Y-%m-%d %H:%M:%S';
has is_local => 0;

# Constructors

sub new {
  my $class = shift;
  my $ts = 0;
  if (ref $class) {
    # Clone
    $ts = $$class;
    $class = ref $class;
    carp "Useless arguments to new (@{[ join ',', @_ ]})" if @_;
  }
  elsif (@_ >= 1) {
    # Pre-generated
    $ts = ref $_[0] ? ${ shift() } : shift;
  }
  return bless \$ts => $class;
}

sub now { my $class = shift; $class = ref $class || $class; $class->new(time) }

sub from_string {
  my ($class, $iso_date, $local) = @_;
  my $sec = 0;
  if ($iso_date =~ /^(\d{4})-(\d{2})-(\d{2})(?:T|\s)(\d{2}):(\d{2}):(\d{2})Z?$/) {
    $sec = ($local || $class->{is_local})
        ? mktime($6, $5, $4, $3, $2 - 1, $1 - 1900)
        : timegm($6, $5, $4, $3, $2 - 1, $1 - 1900);
  }
  return $class->new($sec);
}

# Public methods

sub to_string {
  my ($class, $self, $local) = (undef, shift, undef);
  if (ref $self) {
    # object method
    $class = ref $self;
    $local = shift;
  }
  else {
    # class method
    $class = $self;
    $self = shift;
    $self = \$self unless ref $self;
    $local = shift;
  }
  return ($local || $class->is_local)
      ? strftime $class->format, localtime $$self
      : strftime $class->format, gmtime $$self;
}

1;
__END__


lib/Mojar/Cron/Holiday/UkGov.pm
package Mojar::Cron::Holiday::UkGov;
use Mojo::Base 'Mojar::Cron::Holiday';

our $VERSION = 0.011;

use Mojo::UserAgent;

has ua => sub { Mojo::UserAgent->new(max_redirects => 3) };
has division => 'england-and-wales';
has url => 'https://www.gov.uk/bank-holidays.json';

sub load {
  my ($self, %param) = @_;
  require IO::Socket::SSL;

  my $tx = $self->ua->get($self->url);
  if (my $err = $tx->error) {
    $self->error(sprintf "Failed to fetch holidays (%u)\n%s",
        $err->{advice} // '0', $err->{message} // 'coded error');
    return undef;
  }

  my $loaded = $tx->res->json(sprintf '/%s/events', $self->division);
  return 0 unless @$loaded;

  $self->holidays({}) if $param{reset};
  $self->holiday($_->{date} => 1) for @$loaded;

  return scalar @$loaded;
}

1;
__END__

=head1 NAME

Mojar::Cron::Holiday::UkGov - Feed from gov.uk

=head1 SYNOPSIS

  use Mojar::Cron::Holiday::UkGov;
  my $calendar = Mojar::Cron::Holiday::UkGov->new(division => 'Scotland');
  if ($calendar->load) {
    say 'Whoopee!' if $calendar->holiday($today);
  }

=head1 COPYRIGHT AND LICENCE

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

Copyright (C) 2014, Nic Sandfield.

=head1 SEE ALSO

L<Mojar::Cron::Holiday::Kayaposoft>.


lib/Mojar/Cron/Datetime.pm
package Mojar::Cron::Datetime;
use Mojo::Base -strict;

use Carp qw(carp croak);
use Mojar::ClassShare 'have';
use Mojar::Cron::Util qw(balance life_to_zero normalise_local normalise_utc
    time_to_zero zero_to_time utc_to_ts local_to_ts);
use POSIX 'strftime';
#use Time::Local 'timegm';

our @TimeFields = qw(sec min hour day month year);

# Normal maxima (soft limits)
%Mojar::Cron::Datetime::Max = (
  sec  => 59,
  min  => 59,
  hour => 23,
  day => 30,
  month  => 11,
  weekday => 6
);
@Mojar::Cron::Datetime::Max =
    @Mojar::Cron::Datetime::Max{qw(sec min hour day month weekday)};

# Class attributes
# (not usable on objects)

# Constructors

sub new {
  my $class = shift;
  my $self;
  if (ref $class) {
    # Clone
    $self = [ @$class ];
    $class = ref $class;
    carp "Useless arguments to new (@{[ join ',', @_ ]})" if @_;
  }
  elsif (@_ == 0) {
    # Zero member
    $self = [0,0,0, 0,0,0];
  }
  elsif (@_ == 1) {
    # Pre-generated
    croak "Non-ref argument to new ($self)" unless ref($self = shift);
  }
  else {
    $self = [ @_ ];
  }
  bless $self => $class;
  return $self->normalise;  # Calculate weekday etc
}

sub from_string {
  my ($class, $iso_date) = @_;
  $class = ref $class || $class;
  if ($iso_date
      =~ /^(\d{4})-(\d{2})-(\d{2})(?:T|\s)(\d{2}):(\d{2}):(\d{2})Z?$/) {
    return $class->new(life_to_zero($6, $5, $4, $3, $2, $1));
  }
  croak "Failed to parse datetime string ($iso_date)";
}

sub from_timestamp {
  my ($class, $timestamp, $is_local) = @_;
  $class = ref $class || $class;
  my @parts = $is_local ? localtime $timestamp
                        : gmtime $timestamp;
  return $class->new( time_to_zero @parts );
}

sub now { shift->from_timestamp(time, @_) }

# Public methods

sub copy {
  my ($self, $original) = @_;
  return unless ref $original;
  return $self->clone(@_) unless ref $self;
  @$self = @$original;
  return $self;
}

sub reset_parts {
  my ($self, $end) = @_;
  $$self[$_] = 0 for 0 .. $end;
  return $self;
}

sub weekday {
  my $self = shift;
  $self->normalise;
  return $self->[6];
}

sub normalise {
  my $self = shift;
  my $class = ref $self || $self;
  my @parts = @_ ? @_ : @$self;
  @parts = time_to_zero normalise_utc zero_to_time @parts;
  return @parts if @_;  # operating on argument

  @$self = @parts;  # operating on invocant
  return $self;
}

sub to_timestamp {
  my ($self, $is_local) = @_;
  return $is_local ? local_to_ts zero_to_time @$self
                   : utc_to_ts zero_to_time @$self;
}

sub to_string {
  my $self = shift;
  $self = shift if @_ and ref $_[0];
  return strftime pop || '%Y-%m-%d %H:%M:%S', zero_to_time @$self;
}

1;
__END__

=head1 NAME

Mojar::Cron::Datetime - Lightweight datetime with small footprint

=head1 SYNOPSIS

  use Mojar::Cron::Datetime;
  say Mojar::Cron::Datetime->now->to_string;
  my $d = Mojar::Cron::Datetime->from_string('2001-12-25 00:00:01');
  $d->day($d->day + 14);
  $d->normalise;
  say "$d";

=head1 DESCRIPTION

=head1 CONSTRUCTORS

=head2 C<new>

Construct a datetime from passed arguments.

  $d = Mojar::Cron::Datetime->new;  # zero datetime
  $d = $datetime->new;  # clone
  $d = Mojar::Cron::Datetime->new([00, 00, 20, 26, 06, 112]);
  $d = Mojar::Cron::Datetime->new(00, 00, 20, 26, 06, 112);

The first constructs the zero datetime '1900-01-01 00:00:00'.  The second clones
the value of C<$datetime>.  The third uses the passed value (2012-07-27 21:00:00
London time expressed in UTC).  The fourth does the same but using its own
reference.

=head2 C<now>

  $d = Mojar::Cron::Datetime->now;
  $d = Mojar::Cron::Datetime->now($use_local);
  $d = $d->now;

Constructs a datetime for now.  Uses UTC clock unless passed a true value
(indicating to use local clock).  If called as an object method, ignores the
value of the object, so it gives the same result as the class method.  (Compare
to C<new> which uses the object's value.)

=head2 C<from_string>

  $d = Mojar::Cron::Datetime->from_string('2012-07-27 20:00:00');
  $d = Mojar::Cron::Datetime->from_string('2012-07-28T01:00:00', 1);

Constructs a datetime by parsing an ISO 8601 string.  (The method only supports
the formats shown, where 'T' is optional, and not any of the other 8601
variants.)  Uses UTC clock unless passed a true value (indicating to use local
clock).  Both examples result in the same value if the machine's clock is in
UTC+5.

=head1 METHODS

=head2 C<copy>

  $second = Mojar::Cron::Datetime->new->copy($first);

Copies the constituent values from another datetime object.

=head2 C<normalise>

=head2 C<to_string>

  say "$dt";
  say $dt->to_string;
  say $dt->to_string('%Y-%m-%d %H:%M:%S');
  say Mojar::Cron::Datetime->to_string($dt, '%Y-%m-%d');
  say Mojar::Cron::Datetime->to_string([00,00,00, 25,11,101], '%A');

Stringifies the datetime object using the given format.  The default format is
'%Y-%m-%d %H:%M:%S'.  The first three examples are equivalent.

=head1 SEE ALSO

L<DateTime>, L<Time::Moment>.
